### ISSUE TITLE
Missing y-axis gridline labels when rotated to 90 degrees

### ISSUE THREAD

**Issue by nmstreethran** [SciTools/cartopy#2394]
### Description

Some y-axis gridline labels are missing when rotated to 90 degrees. This doesn't seem to happen when the rotation is 89.9 degrees or less. I only noticed this happening after upgrading Cartopy to 0.23.0.

#### Code to reproduce

```py
import cartopy.crs as ccrs
import matplotlib.pyplot as plt


def plot_example_map(ylabel_rotation):
    fig = plt.figure()
    ax = fig.add_subplot(1, 1, 1, projection=ccrs.PlateCarree())
    ax.set_extent([80, 170, -45, 30], crs=ccrs.PlateCarree())

    ax.stock_img()

    ax.gridlines(
        draw_labels={"bottom": "x", "left": "y"},
        ylabel_style={"rotation": ylabel_rotation},
    )

    ax.coastlines()

    plt.tight_layout()
    plt.show()
```

#### Results

```py
# rotation = 90 degrees
plot_example_map(ylabel_rotation=90)
```

![rotation = 90 degrees](https://github.com/SciTools/cartopy/assets/31223306/078bd6a9-7689-4c41-8a67-7690e9fa5e77)

```py
# rotation = 89.9 degrees
plot_example_map(ylabel_rotation=89.9)
```

![image](https://github.com/SciTools/cartopy/assets/31223306/b5e44ac2-1486-43b6-9844-99f2dae470a9)

<details>
 <summary>Full environment definition</summary>

### Operating system

Linux (Pop!_OS 22.04 LTS x86_64)

### Cartopy version

0.23.0

### pip list

```
cartopy
jupyterlab
```

### Environment

```sh
python3.11 -m venv .venv
source .venv/bin/activate
python -m pip install --upgrade pip setuptools wheel
python -m pip install cartopy jupyterlab
```

</details>

**Comment by rcomer** [comment]
Thanks for the clear report @nmstreethran.  I confirm that I can reproduce this in a script if I replace the `plt.show()` with `plt.savefig('test.png', bbox_inches='tight')`.  In notebooks the `bbox_inches='tight'` is automatic.  If I remove that option in the script, or [disable it in the notebook](https://stackoverflow.com/questions/26714626/how-to-disable-bbox-inches-tight-when-working-with-matplotlib-inline-in-ipytho), then all the labels show.  So we seem to somehow have a bad interaction with `bbox_inches='tight'`.

I think the relevant difference between Cartopy 0.22 and 0.23 is that we changed the default of the _auto_update_ parameter to `True` in #2330.  So the logic that decides where the labels go is now running twice: once when you call `tight_layout` and again when you call `show`.  If you manually set `auto_update=False` in your call to `gridlines` then you get the previous behaviour back.

**Comment by rcomer** [comment]
I added some prints to try to debug this

```diff
diff --git a/lib/cartopy/mpl/gridliner.py b/lib/cartopy/mpl/gridliner.py
index c750a754..6e1e1acb 100644
--- a/lib/cartopy/mpl/gridliner.py
+++ b/lib/cartopy/mpl/gridliner.py
@@ -1277,4 +1277,6 @@ class Label:
         overlapping = self.path.intersects_path(label.path)
         if overlapping:
             self.set_visible(False)
+            print(self.artist, 'overlaps', label.artist)
+            print('overlapping paths:', self.path, label.path, '===', sep='\n')
         return overlapping
```
and got

```
Text(80.00000009, -30.000000000000007, '30°S') overlaps Text(80.00000009, -40.00000000000001, '40°S')
overlapping paths:
Path(array([[ 24.       , 100.1305552],
       [ 24.       , 133.3805552],
       [ 10.       , 133.3805552],
       [ 10.       , 100.1305552],
       [ 24.       , 100.1305552]]), array([ 1,  2,  2,  2, 79], dtype=uint8))
Path(array([[24.        , 42.92314774],
       [24.        , 76.17314774],
       [10.        , 76.17314774],
       [10.        , 42.92314774],
       [24.        , 42.92314774]]), array([ 1,  2,  2,  2, 79], dtype=uint8))
===
Text(80.00000009, -10.000000000000007, '10°S') overlaps Text(80.00000009, -20.000000000000007, '20°S')
overlapping paths:
Path(array([[ 24.        , 214.48287011],
       [ 24.        , 247.85787011],
       [ 10.        , 247.85787011],
       [ 10.        , 214.48287011],
       [ 24.        , 214.48287011]]), array([ 1,  2,  2,  2, 79], dtype=uint8))
Path(array([[ 24.        , 157.33796266],
       [ 24.        , 190.58796266],
       [ 10.        , 190.58796266],
       [ 10.        , 157.33796266],
       [ 24.        , 157.33796266]]), array([ 1,  2,  2,  2, 79], dtype=uint8))
===
Text(80.00000009, -7.105427357601002e-15, '0°') overlaps Text(80.00000009, -20.000000000000007, '20°S')
overlapping paths:
Path(array([[ 24.        , 280.50277757],
       [ 24.        , 296.25277757],
       [ 10.        , 296.25277757],
       [ 10.        , 280.50277757],
       [ 24.        , 280.50277757]]), array([ 1,  2,  2,  2, 79], dtype=uint8))
Path(array([[ 24.        , 157.33796266],
       [ 24.        , 190.58796266],
       [ 10.        , 190.58796266],
       [ 10.        , 157.33796266],
       [ 24.        , 157.33796266]]), array([ 1,  2,  2,  2, 79], dtype=uint8))
===
```
This makes no sense to me.  It's as if the `Path.intersects_path` method is seeing different vertices to what gets printed. :confused:

**Comment by guillermodotn** [comment]
Cartopy uses `intersects_path()` to detect overlapping labels and hide one, but this matplotlib function incorrectly reports non-overlapping rectangles as intersecting when they're perfectly axis-aligned (exactly `rotation=90°`); at `89.9°` the slight tilt avoids the bug.

Swapping it for a simple rectangle overlap check seems to fix it:

https://github.com/SciTools/cartopy/blob/97cfce0027c0fa09a3c673c1b7be98e973146163/lib/cartopy/mpl/gridliner.py#L1289-L1293

```python
# Before
overlapping = self.path.intersects_path(label.path)
# After
overlapping = self.path.get_extents().overlaps(label.path.get_extents())
```

Happy to open a PR with this fix.

**Comment by greglucas** [comment]
I wonder if this has to do with something similar to this issue https://github.com/matplotlib/matplotlib/issues/6076 for the reason they aren't intersecting.

Does checking the bboxes lead to fewer labels being added in general? I'm wondering if the path intersection was chosen so that we weren't checking large rectangles against one another or a tilted geo label's bbox wouldn't be too large and cause excessive padding to make it not intersect the curved boundary. That is all purely speculation though, @guillermodotn please feel free to open PRs if you think you have a fix and other tests don't fail or you have a good reason for choosing something. Always happy to have more contributions fixing things!

**Comment by guillermodotn** [comment]
You're right, https://github.com/matplotlib/matplotlib/issues/6076 seems to be the root cause but it was closed inactive so unlikely to be fixed upstream.

Also right that `Bbox.overlaps` alone gives false positives on tilted geo labels. Went with a bbox pre-check instead: skip `intersects_path()` only when bboxes don't overlap, fall through otherwise.

**Comment by rcomer** [comment]
The workaround mentioned at https://github.com/matplotlib/matplotlib/issues/6076 is to transform the paths to display space before checking intersection.  But I think our paths are already in display space because of this line

https://github.com/SciTools/cartopy/blob/29ebd329d722c867f65031eec663e24ddb2e4a29/lib/cartopy/mpl/gridliner.py#L753-L754

**Comment by guillermodotn** [comment]
Right @rcomer, the paths are already in display space so that workaround doesn't apply. The bbox check circumvents the issue.

Hope I'm not missing any specific case scenario here 🤞

See #2668.
